### DISCUSSION TITLE
How to create SOAP web server using Fastapi ?

### DISCUSSION THREAD

**Discussion by priyatamnano** [fastapi/fastapi#8518] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Union

from fastapi import FastAPI

app = FastAPI()


@app.get("/")
def read_root():
    return {"Hello": "World"}


@app.get("/items/{item_id}")
def read_item(item_id: int, q: Union[str, None] = None):
    return {"item_id": item_id, "q": q}
```

### Description

How to create SOAP web server using Fastapi ?

Note: I know how to create restapi and web server

### Operating System

Linux

### Operating System Details

linux

### FastAPI Version

3.0

### Python Version

3.8

### Additional Context

please help me asap

**Comment by odiseo0** [comment]
Honestly I've never seen someone use FastAPI with SOAP, I guess you could try returning raw xml responses with [Response](https://fastapi.tiangolo.com/advanced/response-directly/), maybe the docs can help you a little and I also think you will need a SOAP client

**Comment by priyatamnano** [comment]
Forget about fastapi ,how can we do in python?

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by chbndrhnns** [comment]
> how can we do in python?

It seems to me this is not the right place to ask such a question.

**Comment by cercide** [comment, marked as answer]
[fastapi-xml](https://github.com/cercide/fastapi-xml) adds xml support to fastapi. Probably that project helps you out.

Here is an example:

```python
from dataclasses import dataclass, field
from fastapi_xml import NonJsonRoute, XmlBody, XmlAppResponse
from fastapi import FastAPI

@dataclass
class HelloWorld:
    message: str = field(metadata={"name": "Message", "type": "Element"})

@dataclass
class Item:
    id: int = field(metadata={"name": "ID", "type": "Attribute"})
    q: str = field(metadata={"name": "q", "type": "Element"})

app = FastAPI(default_response_class=XmlAppResponse)
app.router.route_class = NonJsonRoute

@app.get("/", response_model=HelloWorld)
def read_root() -> HelloWorld:
    return HelloWorld(message="Hello my friend :)")

@app.get("/items/{item_id}", response_model=Item)
def read_item(item_id: int) -> Item:
    return Item(id=item_id, q="Some Item Data")


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="127.0.0.1", port=8000)
```

**Comment by tjoubert** [comment]
Has anyone had any luck with building a soap server with FastAPI?
@cercide, your suggestion on `fastapi-xml` seems good. Perhaps in conjunction with a custom middleware (https://fastapi.tiangolo.com/tutorial/middleware/ )? 
I welcome any suggestion or tips. I am working with an integration where **I have to** build a soap server.

**Comment by Kunika0503** [comment]
has anyone tried using the fastapi-soap library? if yes how can we use it for a custom webservice?

**Comment by yinziyan1206** [comment]
Maybe we should find how to create soap server using python first
